src/Catalyst.Modules.UPnP/PortMapper.cs

```diff
@@ -0,0 +1,154 @@
+﻿#region LICENSE
+
+/**
+* Copyright (c) 2019 Catalyst Network
+*
+* This file is part of Catalyst.Node <https://github.com/catalyst-network/Catalyst.Node>
+*
+* Catalyst.Node is free software: you can redistribute it and/or modify
+* it under the terms of the GNU General Public License as published by
+* the Free Software Foundation, either version 2 of the License, or
+* (at your option) any later version.
+*
+* Catalyst.Node is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with Catalyst.Node. If not, see <https://www.gnu.org/licenses/>.
+*/
+#endregion
+
+using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Mono.Nat;
+using Serilog;
+
+namespace Catalyst.Modules.UPnP
+{
+    public sealed class PortMapper
+    {
+        private readonly SemaphoreSlim _locker = new SemaphoreSlim(1, 1);
+        private readonly INatUtilityProvider _natUtilityProvider;
+        private readonly ILogger _logger;
+
+        public PortMapper(INatUtilityProvider natUtilityProvider, ILogger logger)
+        {
+            _natUtilityProvider = natUtilityProvider;
+            _logger = logger;
+        }
+
+        private async Task<PortMapperConstants.Result> PerformEventOnDeviceFound(int timeoutInSeconds, CancellationToken token, Action<INatDevice> onDeviceFound)
+        {
+            void DeviceFoundFunc(object o, DeviceEventArgs args) => onDeviceFound(args.Device);
+            _natUtilityProvider.DeviceFound += DeviceFoundFunc;
+            
+            _natUtilityProvider.StartDiscovery();
+            _logger.Information(PortMapperConstants.StartedSearching);
+
+            var result = await DelayUntilTimeoutOrMappingTaskEnds(timeoutInSeconds, token);
+
+            _natUtilityProvider.DeviceFound -= DeviceFoundFunc;
+            
+            _natUtilityProvider.StopDiscovery();
+            _logger.Information(PortMapperConstants.StoppedSearching);
+            
+            if(result==PortMapperConstants.Result.Timeout){_logger.Information(PortMapperConstants.CouldNotCommunicateWithRouter);}
+
+            return result;
+        }
+
+        private static async Task<PortMapperConstants.Result> DelayUntilTimeoutOrMappingTaskEnds(int timeoutInSeconds, CancellationToken token)
+        {
+            try
+            {
+                await Task.Delay(TimeSpan.FromSeconds(timeoutInSeconds), token);
+            }
+            catch(TaskCanceledException)
+            {
+                return PortMapperConstants.Result.TaskFinished;
```

Comment:
Can this be return PortMapperConstants.Result.TaskFinishedWithError;?

Reply:
I've not changed this - Delay(TimeSpan, CancellationToken) throws an exception when the token is cancelled, however this cancellation is expected behaviour for this code - the task has finished normally. To avoid having to throw and catch an exception here I could implement a while loop and await for small chunks of time with Delay(TimeSpan) then check the token repeatedly, but what I've done above seemed cleaner to me.